### ISSUE TITLE
Proof progresses upon reload

### ISSUE THREAD

**Issue by palinatolmach** [runtimeverification/k#4190]
While trying to add a test for failure_info to [test_failure_info_recomputed_on_proof_reinit](https://github.com/runtimeverification/pyk/blob/a442ceaebeef0872de26749d1513fc37d9675206/src/tests/integration/proof/test_imp.py#L1070), I noticed strange behavior where the proof is written to and then read from disk.

The claim, which terminates early, has 1 terminal and 1 pending node when it's written to the disk, but after it's reloaded and is used to initialize APRProver, it has 2 terminal and 0 pending node — in other words, it looks like the pending node gets executed during APRProver initialization:

```py
        # reload proof from disk
        proof = APRProof.read_proof_data(proof_dir, proof_id)
        prover = APRProver(proof, kcfg_explore=kcfg_explore)
```

To reproduce, you can comment out skipping the [test](https://github.com/runtimeverification/pyk/blob/a442ceaebeef0872de26749d1513fc37d9675206/src/tests/integration/proof/test_imp.py#L733) that fails early with `fail_fast` in  [test_failure_info_recomputed_on_proof_reinit](https://github.com/runtimeverification/pyk/blob/a442ceaebeef0872de26749d1513fc37d9675206/src/tests/integration/proof/test_imp.py#L1070):
https://github.com/runtimeverification/pyk/blob/a442ceaebeef0872de26749d1513fc37d9675206/src/tests/integration/proof/test_imp.py#L1084-L1085

**Comment by ehildenb** [comment]
It's because of this line here: https://github.com/runtimeverification/pyk/blob/e243908892fa1d959743cd24dd9ff4a7b9efad00/src/pyk/proof/reachability.py#L669

Shortest thing we can do is delete that line. But ideally we would be checking if a node is terminal every time we insert it into the KCFG. We may also be able to

**Comment by yale-vinson** [comment]
@palinatolmach Given that this is a Bug and categorized as MH, I am inclined to move it to "Next Up", but could be convinced that to "Backlog" it. Thoughts?

**Comment by palinatolmach** [comment]
If I understand correctly, it will be addressed in scope of https://github.com/runtimeverification/pyk/issues/965 that is In Progress and that @nwatson22 is assigned to. @nwatson22 shall we move this issue to In Progress as well?

**Comment by yale-vinson** [comment]
Blocked by runtimeverification/pyk#965 and will open back up once that comes up from the pyk backlog. Moving to Blocked for now.

**Comment by nwatson22** [comment]
runtimeverification/pyk#965 is merged, but we ended up merging a more limited change that doesn't do checking for the attributes whenever they are added. I'm looking at a way to do this, but I am having trouble making it fit within the `Prover`/`Proof` separation framework we are using. The attribute we are concerned with here, `terminal`, is determined by `is_terminal`, held by the `KCFGSemantics`, which is held by `KCFGExplore`, which is held by `APRProver`. Doing something whenever a node is added to a proof sounds like something that should be done in the proof class, but `KCFGSemantics` is supposed to provide hooks that dictate proving logic, which makes sense to be in the purview of `Prover`.

Should we move `KCFGSemantics` to be owned instead by the `Proof`s, try to pass in the `is_terminal` logic to the proof whenever the Prover handles it, or is there a better solution? @ehildenb @palinatolmach

**Comment by palinatolmach** [comment]
Related: https://github.com/runtimeverification/k/issues/4213

**Comment by nwatson22** [comment]
The simple solution to this, to simply not check all terminals on reload, is already implemented (`check_all_terminals` was removed in #4280). #4327 adds a test to demonstrate this and also fixes and re-enables the `fail-fast` test in `test_failure_info_recomputed_on_proof_reinit`.

**Comment by palinatolmach** [comment]
Closing as completed in https://github.com/runtimeverification/k/pull/4280 and https://github.com/runtimeverification/k/pull/4327.
